File: src/main/java/net/casualuhc/uhcmod/utils/screen/SpectatorScreen.java
package net.casualuhc.uhcmod.utils.screen;

import net.casualuhc.uhcmod.UHCMod;
import net.casualuhc.uhcmod.managers.TeamManager;
import net.casualuhc.uhcmod.utils.uhc.Config;
import net.casualuhc.uhcmod.utils.uhc.ItemUtils;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.entity.player.PlayerInventory;
import net.minecraft.item.ItemStack;
import net.minecraft.item.Items;
import net.minecraft.item.SkullItem;
import net.minecraft.nbt.NbtCompound;
import net.minecraft.scoreboard.Team;
import net.minecraft.screen.SimpleNamedScreenHandlerFactory;
import net.minecraft.screen.slot.SlotActionType;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.text.Text;
import net.minecraft.util.math.Vec3d;
import net.minecraft.world.GameMode;

import java.util.List;

public class SpectatorScreen extends CustomScreen {
	private final boolean spectators;
	private final int page;

	public SpectatorScreen(PlayerInventory playerInventory, int syncId, int page, boolean spectators) {
		super(playerInventory, syncId, 6);
		this.page = page;
		this.spectators = spectators;

		if (this.spectators) {
			this.createSpectatorsList();
		} else {
			this.createTeamList();
		}
	}

	@Override
	public void onSlotClick(int slotIndex, int button, SlotActionType actionType, PlayerEntity player) {
		if (slotIndex == 0 && this.page > 0) {
			player.openHandledScreen(SpectatorScreen.createScreenFactory(this.page - 1, this.spectators));
			return;
		}
		if (slotIndex == 7) {
			Vec3d spawn = Config.CURRENT_EVENT.getLobbySpawnPos();
			((ServerPlayerEntity) player).teleport(UHCMod.SERVER.getOverworld(), spawn.getX(), spawn.getY(), spawn.getZ(), 0, 0);
			((ServerPlayerEntity) player).interactionManager.changeGameMode(GameMode.ADVENTURE);
			return;
		}
		if (slotIndex == 8) {
			player.openHandledScreen(SpectatorScreen.createScreenFactory(this.page + 1, this.spectators));
			return;
		}
		if (slotIndex < 9 || slotIndex > 53 || (!this.spectators && slotIndex % 9 == 0)) {
			return;
		}

		ItemStack clickedStack = this.slots.get(slotIndex).getStack();
		if (clickedStack.isEmpty()) {
			return;
		}

		NbtCompound compound = clickedStack.getNbt();
		NbtCompound skullOwner;
		if (compound != null && (skullOwner = compound.getCompound(SkullItem.SKULL_OWNER_KEY)) != null) {
			String playerName = skullOwner.getString("Name");
			ServerPlayerEntity target = UHCMod.SERVER.getPlayerManager().getPlayer(playerName);
			if (target == null) {
				player.sendMessage(Text.translatable("uhc.spectator.notOnline", playerName), false);
				return;
			}
			((ServerPlayerEntity) player).teleport(target.getWorld(), target.getX(), target.getY(), target.getZ(), target.getYaw(), target.getPitch());
		}
	}

	private void createTeamList() {
		this.modifyInventory(inventory -> {
			List<Team> teamList = UHCMod.SERVER.getScoreboard().getTeams()
				.stream()
				.filter(t -> !TeamManager.shouldIgnoreTeam(t) && !t.getPlayerList().isEmpty() && t.getPlayerList().stream().anyMatch(p -> UHCMod.SERVER.getPlayerManager().getPlayer(p) != null))
				.skip(5L * this.page)
				.toList();

			for (int i = 1; i <= 5; i++) {
				if (teamList.size() <= (i - 1)) {
					break;
				}
				int startSlot = i * 9;
				Team team = teamList.get(i - 1);
				inventory.setStack(startSlot, getDisplay(team));

				startSlot += 1;
				List<ItemStack> itemStacks = team.getPlayerList()
					.stream()
					.filter(p -> UHCMod.SERVER.getPlayerManager().getPlayer(p) != null)
					.limit(7)
					.map(p -> ItemUtils.generatePlayerHead(p).setCustomName(Text.literal(p).styled(s -> s.withItalic(false))))
					.toList();
				int options = itemStacks.size();
				for (int j = 0; j < options; j++) {
					inventory.setStack(startSlot + j, itemStacks.get(j));
				}
			}

			for (int i = 1; i < 7; i++) {
				inventory.setStack(i, Items.GRAY_STAINED_GLASS.getDefaultStack().setCustomName(Text.literal("")));
			}

			inventory.setStack(0, ItemUtils.literalNamed(Items.RED_STAINED_GLASS, "Previous"));
			inventory.setStack(7, ItemUtils.literalNamed(Items.BLUE_STAINED_GLASS, "Return to Bleachers"));
			inventory.setStack(8, ItemUtils.literalNamed(Items.GREEN_STAINED_GLASS, "Next"));
		});
	}

	private void createSpectatorsList() {
		this.modifyInventory(inventory -> {
			List<ItemStack> itemStacks = UHCMod.SERVER.getScoreboard().getTeams()
				.stream()
				.filter(TeamManager::shouldIgnoreTeam)
				.flatMap(t -> t.getPlayerList().stream())
				.filter(p -> UHCMod.SERVER.getPlayerManager().getPlayer(p) != null)
				.skip(45L * this.page)
				.limit(45)
				.map(p -> ItemUtils.generatePlayerHead(p).setCustomName(Text.literal(p).styled(s -> s.withItalic(false))))
				.toList();
			for (int i = 0; i < itemStacks.size(); i++) {
				ItemStack stack = itemStacks.get(i);
				inventory.setStack(i + 9, stack);
			}

			for (int i = 1; i < 7; i++) {
				inventory.setStack(i, Items.GRAY_STAINED_GLASS.getDefaultStack().setCustomName(Text.literal("")));
			}

			inventory.setStack(0, ItemUtils.literalNamed(Items.RED_STAINED_GLASS, "Previous"));
			inventory.setStack(7, ItemUtils.literalNamed(Items.BLUE_STAINED_GLASS, "Teams"));
			inventory.setStack(8, ItemUtils.literalNamed(Items.GREEN_STAINED_GLASS, "Next"));
		});
	}

	private static ItemStack getDisplay(Team team) {
		String texture = switch (team.getColor()) {
			case BLACK -> ItemUtils.BLACK;
			case DARK_BLUE -> ItemUtils.DARK_BLUE;
			case DARK_GREEN -> ItemUtils.DARK_GREEN;
			case DARK_AQUA -> ItemUtils.DARK_AQUA;
			case DARK_RED -> ItemUtils.DARK_RED;
			case DARK_PURPLE -> ItemUtils.DARK_PURPLE;
			case GOLD -> ItemUtils.GOLD;
			case GRAY -> ItemUtils.GRAY;
			case DARK_GRAY -> ItemUtils.DARK_GRAY;
			case BLUE -> ItemUtils.BLUE;
			case GREEN -> ItemUtils.GREEN;
			case AQUA -> ItemUtils.AQUA;
			case RED -> ItemUtils.RED;
			case LIGHT_PURPLE -> ItemUtils.LIGHT_PURPLE;
			case YELLOW -> ItemUtils.YELLOW;
			default -> ItemUtils.WHITE;
		};
		return ItemUtils.generatePlayerHead("Dummy", texture).setCustomName(Text.literal(team.getName()).styled(s -> s.withItalic(false)));
	}

	public static SimpleNamedScreenHandlerFactory createScreenFactory(int page, boolean spectators) {
		if (page < 0) {
			return null;
		}
		return new SimpleNamedScreenHandlerFactory((syncId, inv, player) -> {
			return new SpectatorScreen(inv, syncId, page, spectators);
		}, Text.translatable("uhc.spectator.screen"));
	}
}

File: src/main/kotlin/net/casualuhc/uhcmod/screen/SpectatorScreen.kt
package net.casualuhc.uhcmod.screen

import net.casualuhc.arcade.Arcade
import net.casualuhc.arcade.utils.ComponentUtils.unItalicise
import net.casualuhc.arcade.utils.ItemUtils
import net.casualuhc.arcade.utils.ItemUtils.literalNamed
import net.casualuhc.arcade.utils.TeamUtils.getServerPlayers
import net.casualuhc.uhcmod.extensions.TeamFlag.Ignored
import net.casualuhc.uhcmod.extensions.TeamFlagsExtension.Companion.flags
import net.casualuhc.uhcmod.util.HeadUtils
import net.casualuhc.uhcmod.util.Texts
import net.minecraft.ChatFormatting
import net.minecraft.network.chat.Component
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.SimpleMenuProvider
import net.minecraft.world.entity.player.Inventory
import net.minecraft.world.entity.player.Player
import net.minecraft.world.inventory.ClickType
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.Items
import net.minecraft.world.item.PlayerHeadItem
import net.minecraft.world.scores.Team

class SpectatorScreen(
    inventory: Inventory,
    syncId: Int,
    private val page: Int,
    private val spectators: Boolean
): CustomScreen(inventory, syncId, 6) {
    init {
        if (this.spectators) {
            createSpectatorsList()
        } else {
            createTeamList()
        }
    }

    override fun clicked(slotId: Int, button: Int, clickType: ClickType, player: Player) {
        if (slotId == 0 && page > 0) {
            player.openMenu(createScreenFactory(page - 1, spectators))
            return
        }
        if (slotId == 7) {
            player.openMenu(createScreenFactory(0, !spectators))
            return
        }
        if (slotId == 8) {
            player.openMenu(createScreenFactory(page + 1, spectators))
            return
        }
        if (slotId < 9 || slotId > 53 || !spectators && slotId % 9 == 0) {
            return
        }
        val clickedStack = this.slots[slotId].item
        if (clickedStack.isEmpty) {
            return
        }

        val compound = clickedStack.tag ?: return
        val owner = compound.getCompound(PlayerHeadItem.TAG_SKULL_OWNER) ?: return
        val playerName: String = owner.getString("Name")
        val target = Arcade.server.playerList.getPlayerByName(playerName)
        if (target == null) {
            player.sendSystemMessage(Texts.SPECTATOR_NOT_ONLINE.generate(playerName))
            return
        }
        (player as ServerPlayer).teleportTo(
            target.serverLevel(),
            target.x,
            target.y,
            target.z,
            target.xRot,
            target.yRot
        )
    }

    private fun createTeamList() {
        this.modifyInventory { inventory ->
            val teamList = Arcade.server.scoreboard.playerTeams
                .stream()
                .filter { !it.flags.has(Ignored) && it.getServerPlayers().isNotEmpty() }
                .skip(5L * page)
                .toList()

            for (i in 1..5) {
                if (teamList.size <= i - 1) {
                    break
                }
                var startSlot = i * 9
                val team = teamList[i - 1]
                inventory.setItem(startSlot, getDisplay(team))
                startSlot += 1
                val itemStacks: List<ItemStack> = team.getServerPlayers()
                    .stream()
                    .limit(7)
                    .map { HeadUtils.generateHead(it).setHoverName(Component.literal(it.scoreboardName).unItalicise()) }
                    .toList()
                val options = itemStacks.size
                for (j in 0 until options) {
                    inventory.setItem(startSlot + j, itemStacks[j])
                }
            }
            for (i in 1..6) {
                inventory.setItem(i, Items.GRAY_STAINED_GLASS.literalNamed(""))
            }

            inventory.setItem(0, Items.RED_STAINED_GLASS.defaultInstance.setHoverName(Texts.SPECTATOR_PREVIOUS))
            inventory.setItem(7, Items.BLUE_STAINED_GLASS.defaultInstance.setHoverName(Texts.SPECTATOR_SPECTATORS))
            inventory.setItem(8, Items.GREEN_STAINED_GLASS.defaultInstance.setHoverName(Texts.SPECTATOR_NEXT))
        }
    }

    private fun createSpectatorsList() {
        this.modifyInventory { inventory ->
            val itemStacks: List<ItemStack> = Arcade.server.scoreboard.playerTeams
                .stream()
                .filter { it.flags.has(Ignored) }
                .flatMap { it.getServerPlayers().stream() }
                .skip(45L * page)
                .limit(45)
                .map { HeadUtils.generateHead(it).setHoverName(Component.literal(it.scoreboardName).unItalicise()) }
                .toList()
            for (i in itemStacks.indices) {
                val stack: ItemStack = itemStacks[i]
                inventory.setItem(i + 9, stack)
            }
            for (i in 1..6) {
                inventory.setItem(i, Items.GRAY_STAINED_GLASS.literalNamed(""))
            }

            inventory.setItem(0, Items.RED_STAINED_GLASS.defaultInstance.setHoverName(Texts.SPECTATOR_PREVIOUS))
            inventory.setItem(7, Items.BLUE_STAINED_GLASS.defaultInstance.setHoverName(Texts.SPECTATOR_TEAMS))
            inventory.setItem(8, Items.GREEN_STAINED_GLASS.defaultInstance.setHoverName(Texts.SPECTATOR_NEXT))
        }
    }

    companion object {
        private fun getDisplay(team: Team): ItemStack {
            val texture = when (team.color) {
                ChatFormatting.BLACK -> HeadUtils.BLACK
                ChatFormatting.DARK_BLUE -> HeadUtils.DARK_BLUE
                ChatFormatting.DARK_GREEN -> HeadUtils.DARK_GREEN
                ChatFormatting.DARK_AQUA -> HeadUtils.DARK_AQUA
                ChatFormatting.DARK_RED -> HeadUtils.DARK_RED
                ChatFormatting.DARK_PURPLE -> HeadUtils.DARK_PURPLE
                ChatFormatting.GOLD -> HeadUtils.GOLD
                ChatFormatting.GRAY -> HeadUtils.GRAY
                ChatFormatting.DARK_GRAY -> HeadUtils.DARK_GRAY
                ChatFormatting.BLUE -> HeadUtils.BLUE
                ChatFormatting.GREEN -> HeadUtils.GREEN
                ChatFormatting.AQUA -> HeadUtils.AQUA
                ChatFormatting.RED -> HeadUtils.RED
                ChatFormatting.LIGHT_PURPLE -> HeadUtils.LIGHT_PURPLE
                ChatFormatting.YELLOW -> HeadUtils.YELLOW
                else -> HeadUtils.WHITE
            }
            return ItemUtils.generatePlayerHead("Dummy", texture).setHoverName(Component.literal(team.name).unItalicise())
        }

        fun createScreenFactory(page: Int, spectators: Boolean): SimpleMenuProvider? {
            return if (page < 0) {
                null
            } else SimpleMenuProvider(
                { syncId, inv, player -> SpectatorScreen(inv, syncId, page, spectators) },
                Texts.SPECTATOR_SCREEN
            )
        }
    }
}
